src/nimble.nim

```diff
@@ -557,48 +366,69 @@ proc downloadDependency(name: string, dep: LockFileDep, options: Options):
     downloadDir: downloadDir,
     vcsRevision: vcsRevision)
 
-proc installDependency(pkgInfo: PackageInfo, downloadInfo: DownloadInfo,
-                       options: Options): PackageInfo =
+proc installLockedDependency(parentInfo: PackageInfo, downloadInfo: DownloadInfo,
+                             paths: seq[string], options: Options): PackageInfo =
   ## Installs an already downloaded dependency of the package `pkgInfo`.
-  let (_, newlyInstalledPkgInfo) = installFromDir(
+  var newlyInstalledPkgInfo = getPkgInfo(downloadInfo.downloadDir, options)
+  installFromDir(
     downloadInfo.downloadDir,
-    downloadInfo.version,
     options,
+    paths,
     downloadInfo.url,
-    first = false,
-    fromLockFile = true,
-    downloadInfo.vcsRevision)
+    newlyInstalledPkgInfo
+    )
 
   downloadInfo.downloadDir.removeDir
 
   for depDepName in downloadInfo.dependency.dependencies:
-    let depDep = pkgInfo.lockedDeps[depDepName]
+    let depDep = parentInfo.lockedDeps[depDepName]
     let revDep = (name: depDepName, version: depDep.version,
                   checksum: depDep.checksums.sha1)
     options.nimbleData.addRevDep(revDep, newlyInstalledPkgInfo)
 
   return newlyInstalledPkgInfo
 
+proc getLockedDependency(name: string, dep: LockFileDep, options: Options):
+    PackageInfo =
+  ## Returns a `PackageInfo` for an already installed dependency from the
+  ## lock file.
+  let depDirName = getDependencyDir(name, dep, options)
+  let nimbleFilePath = findNimbleFile(depDirName, false)
+  getInstalledPackageMin(depDirName, nimbleFilePath).toFullInfo(options)
+
+proc isLockedInstalled(name: string, dep: LockFileDep, options: Options): bool =
+  ## Checks whether a dependency from the lock file is already installed.
+  fileExists(getDependencyDir(name, dep, options) / packageMetaDataFileName)
+
 proc processLockedDependencies(pkgInfo: PackageInfo, options: Options):
-    HashSet[PackageInfo] =
-  # Returns a hash set with `PackageInfo` of all packages from the lock file of
+    seq[PackageInfo] =
+  # Returns a seq with `PackageInfo` of all packages from the lock file of
   # the package `pkgInfo` by getting the info for develop mode dependencies from
   # their local file system directories and other packages from the Nimble
   # cache. If a package with required checksum is missing from the local cache
   # installs it by downloading it from its repository.
 
-  let developModeDeps = getDevelopDependencies(pkgInfo, options)
+  let
+    depsGraph = toSeq(pkgInfo.lockedDeps.pairs).map(pair => (pair[0], pair[1].dependencies))
+    (order, _) = topologicalSort(depsGraph.toOrderedTable())
+    developModeDeps = getDevelopDependencies(pkgInfo, options)
```

Comment:
The current master version relies on json object key ordering to install locked deps in the right order, but json doesn't give this guarantee afaik
I added a second sort here, to be safe. See also comment on `lock(`